src/install.rs

```diff
@@ -24,6 +24,65 @@ use crate::download::*;
 use crate::errors::*;
 use crate::source::*;
 
+/// Integrity verification hash for an Ignition config.
+#[derive(Debug)]
+pub enum IgnitionHash {
+    /// SHA-512 digest.
+    Sha512(Vec<u8>),
+}
+
+impl IgnitionHash {
+    /// Try to parse an hash-digest argument.
+    ///
+    /// This expects an input value following the `ignition.config.verification.hash`
+    /// spec, i.e. `<type>-<value>` format.
+    pub fn try_parse(input: &str) -> Result<Self> {
+        let parts: Vec<_> = input.splitn(2, '-').collect();
+        if parts.len() != 2 {
+            bail!("failed to detect hash-type and digest in '{}'", input);
+        }
+        let (hash_kind, hex_digest) = (parts[0], parts[1]);
+
+        let hash = match hash_kind {
+            "sha512" => {
+                let digest = hex::decode(hex_digest).chain_err(|| "decoding hex digest")?;
+                ensure!(
+                    digest.len().saturating_mul(8) == 512,
+                    "wrong digest length ({})",
+                    digest.len().saturating_mul(8)
+                );
+                IgnitionHash::Sha512(digest)
+            }
+            x => bail!("unknown hash type '{}'", x),
+        };
+
+        Ok(hash)
+    }
+
+    /// Digest and validate input data.
+    pub fn validate<R>(&self, input: &mut R) -> Result<()>
+    where
+        R: Read,
```

Comment:
(This sent me on a very long read and ticket-hopping journey on existential types, but I'm very glad you mentioned it).

So, to unwrap this a bit, you are suggesting to more broadly start adopting [anonymous type parameters](https://doc.rust-lang.org/reference/types/impl-trait.html#anonymous-type-parameters) for arguments, instead of [generics with trait bounds](https://doc.rust-lang.org/reference/items/generics.html), right?
I'm on board with that, even if the former is a somehow recent thing which I've rarely seen in codebases.

Reply:
> So, to unwrap this a bit, you are suggesting to more broadly start adopting anonymous type parameters for arguments, instead of generics with trait bounds, right?

Yes, exactly. Generics still have their place of course if e.g. you have multiple trait bounds or you want it to be part of a public API. But otherwise for internal functions, I find `impl Trait` nicer overall. That's my personal opinion though. :)